### DISCUSSION TITLE
A better way to specify fixtures

### DISCUSSION THREAD

**Discussion by binford2k** [puppetlabs/pdk#1049] [Feature Request]
Right now direct dependencies are specified in `metadata.json` and all modules required for testing, including fully expanded transitive dependencies, are specified in `.fixtures.yml`. This is often incredibly repetitive and I would love it if there were a way to resolve fixtures automatically, sort of the way `bundle install` will install gems and create a lockfile.

Eventually, this pattern could also be used to do things like resolve `Puppetfile` dependencies too.

**Comment by DavidS** [comment]
One of the ideas we've been bouncing around litmus is create a test-specific bolt project with all the stuff needed for testing (inventory, provisioning plan, dependencies, etc). That would allow us to use the [`bolt-project.yaml`](https://puppet.com/docs/bolt/latest/bolt_installing_modules.html#manually-specify-modules-in-a-bolt-project) (maybe even dynamically generated from `metadata.json`) to resolve dependencies without having to manually maintain them.

**Comment by logicminds** [comment]
This has been a long running discussion with several attempts to resolve.

* https://github.com/puppetlabs/puppetlabs_spec_helper/pull/291
* https://github.com/puppetlabs/puppetlabs_spec_helper/pull/109

One of the main problems is what happens when all the modules are coming from VCS instead of the forge for companies that are air-gapped, private modules or have strict security requirements.

## Solution
Personally, this solution needs to be:

1. backwards compatible
2. easy to use


What comes to mind is the Puppetfile syntax.
1.  Existing tools already work with it
2. Everyone knows it
3. Is ruby code, so technically can do fun things inside the puppetfile

If a Puppetfile exist, use Puppetfile instead in place of fixtures. 
Have a property in metadata that enables the usage of Puppetfile for dependency management

```
"resolve_dependencies_with_puppetfile": true
"dependencies": [],
  
```

In order to populate the dependency list from the puppetfile, the `pdk validate dependencies` command should update the list automatically.  BTW, having a dependency checker in the PDK would be amazing for those times you forget to update the metadata or just lazy.

For the .fixtures.yml file.  The puppetlabs_spec_helper would need to convert the puppetfile to fixtures internally.  But only if the resolve_dependencies_with_puppetfile is set to true.

This all seems doable IMHO.   

1. Refactor puppetlabs_spec_helper to work with Puppetfile 
2. Have puppetlabs_spec_helper seek out the new flag in the metadata.json file
3. Have `pdk validate dependencies` auto populate metadata.json dependencies based on Puppetfile
4. Optional, Puppetfile needs a way to specify dev dependencies like nwops-debug and others.

Hopefully, all of this makes sense.
